controllers/ocirepository_controller.go

```diff
@@ -0,0 +1,484 @@
+		return
+	}
+	objRef, err := reference.GetReference(r.Scheme, &repository)
+	if err != nil {
+		log.Error(err, "unable to record readiness metric")
+		return
+	}
+	if rc := apimeta.FindStatusCondition(repository.Status.Conditions, meta.ReadyCondition); rc != nil {
+		r.MetricsRecorder.RecordCondition(*objRef, *rc, !repository.DeletionTimestamp.IsZero())
+	} else {
+		r.MetricsRecorder.RecordCondition(*objRef, metav1.Condition{
+			Type:   meta.ReadyCondition,
+			Status: metav1.ConditionUnknown,
+		}, !repository.DeletionTimestamp.IsZero())
+	}
+}
+
+func (r *OCIRepositoryReconciler) recordSuspension(ctx context.Context, repository sourcev1.OCIRepository) {
+	if r.MetricsRecorder == nil {
+		return
+	}
+	log := logr.FromContext(ctx)
+
+	objRef, err := reference.GetReference(r.Scheme, &repository)
+	if err != nil {
+		log.Error(err, "unable to record suspended metric")
+		return
+	}
+
+	if !repository.DeletionTimestamp.IsZero() {
+		r.MetricsRecorder.RecordSuspend(*objRef, false)
+	} else {
+		r.MetricsRecorder.RecordSuspend(*objRef, repository.Spec.Suspend)
+	}
+}
+
+// resetStatus returns a modified v1beta1.OCIRepository and a boolean indicating
+// if the status field has been reset.
+func (r *OCIRepositoryReconciler) resetStatus(repository sourcev1.OCIRepository) (sourcev1.OCIRepository, bool) {
+	// We do not have an artifact, or it does no longer exist
+	if repository.GetArtifact() == nil || !r.Storage.ArtifactExist(*repository.GetArtifact()) {
+		repository = sourcev1.OCIRepositoryProgressing(repository)
+		repository.Status.Artifact = nil
+		return repository, true
+	}
+	if repository.Generation != repository.Status.ObservedGeneration {
+		return sourcev1.OCIRepositoryProgressing(repository), true
+	}
+	return repository, false
+}
+
+func (r *OCIRepositoryReconciler) reference(ctx context.Context, keychain authn.Keychain, repository sourcev1.OCIRepository) (name.Reference, error) {
+	url := repository.Spec.URL
+
+	ref := repository.Spec.Reference
+	if ref == nil {
+		return name.ParseReference(fmt.Sprintf("%s:latest", url))
+	}
+
+	if ref.Digest != "" {
+		return name.ParseReference(fmt.Sprintf("%s@%s", url, ref.Digest))
+	}
+
+	if ref.SemVer != "" {
+		repo, err := name.NewRepository(url)
+		if err != nil {
+			return nil, err
+		}
+
+		tags, err := remote.List(repo, remote.WithAuthFromKeychain(keychain), remote.WithContext(ctx))
+		if err != nil {
+			return nil, err
+		}
+
+		c, err := semver.NewConstraint(ref.SemVer)
+		if err != nil {
+			return nil, err
+		}
+
+		var candidates []*semver.Version
+		for _, t := range tags {
+				v, err := semver.NewVersion(t)
+				if err != nil {
+					continue
+				}
+
+				if c != nil && !c.Check(v) {
+					continue
+				}
+
+				candidates = append(candidates, v)
+		}
+
+		if len(candidates) == 0 {
+			return nil, fmt.Errorf("no matching tags were found")
+		}
+
+		sort.Sort(sort.Reverse(semver.Collection(candidates)))
+		return name.ParseReference(fmt.Sprintf("%s:%s", url, candidates[0]))
```

Comment:
The `semver.NewVersion` strips the `v` prefix, so we can't use it to build the URL, please use `candidates[0].Original()`.